keps/sig-security/2763-ambient-capabilities/README.md

```diff
@@ -0,0 +1,902 @@
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+- We propose updating the kubernetes core and CRI API's to allow users to configure adding capabilities to the ambient set. This would allow users to run non-root containers without having to apply file capabilities to the binary during image build time. Details on how and where these changes will be made are explained in the [Changes to kubernetes API (<a href="https://pkg.go.dev/k8s.io/api/core/v1">https://pkg.go.dev/k8s.io/api/core/v1</a>)](#changes-to-kubernetes-api-httpspkggodevk8sioapicorev1) and [Changes to runtime API (<a href="https://pkg.go.dev/k8s.io/cri-api/pkg/apis/runtime/v1alpha2">https://pkg.go.dev/k8s.io/cri-api/pkg/apis/runtime/v1alpha2</a>)](#changes-to-runtime-api-httpspkggodevk8siocri-apipkgapisruntimev1alpha2) sections below.
+
+- We also propose that container runtimes like `containerd` and `CRI-O` be updated to account for these changes to the CRI apis and add the requested capabilities to the ambient set when they create the containers. Details on how and where these changes will be made are explained in the [Changes to containerd/containerd (<a href="https://github.com/containerd/containerd">https://github.com/containerd/containerd</a>)](#changes-to-containerdcontainerd-httpsgithubcomcontainerdcontainerd) section below.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+As a security conscious user I should be able to run my container as non-root and add the minimum set of capabilities required for it to function even when I do not control how the image is built.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Changes to kubernetes API (https://pkg.go.dev/k8s.io/api/core/v1)
+
+<<[UNRESOLVED should we reuse add instead of adding new field]>>
```

Comment:
based on this note https://github.com/kubernetes/kubernetes/issues/56374#issuecomment-395814330
I don't think we should reuse add. I think having an additional slice in the Capabilities struct that is add-only is a clean addition

Reply:
Here is my two cents about the k8s API design. (Talked with Vinayak offline already)

In one end of the spectrum, k8s API can expose the exact interfaces of the layer below it. In this case, it is [Linux capabilities API](https://man7.org/linux/man-pages/man7/capabilities.7.html). It means k8s API can just allow user directly interact with permitted, inherited, effective, bounding and ambient sets. This way gives user 100% control and flexibility as what Linux allows. This also means k8s API need to adapt with Linux capabilities API. For example, a new thing in Linux capabilities results in a new thing in k8s API.

In the other end, k8s API defines and abstracts its own way of interacting with capabilities. So k8s API can be simpler, at the cost of flexibility. And it should hide details of the underlying OS. The original design of `securityContext.capabilities.add` and `drop` are pretty concise. But it is unclear if it supports both root and non-root. It seems to support but from [this k8s doc](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#set-capabilities-for-a-container), it doesn't support non root.

> Linux Capabilities: Give a process some privileges, but not all the privileges of the root user.

Reply:
At this moment, we need to be backward-compatible, which is the only reason that I agree we need a new field. With the new field, I propose to **clarify** the spec, based on [kubernetes/kubernetes#56374 (comment)].

- `add`: if root user, add capabilities; if non-root, behavior is unspecified
- `drop`: if root user, drop capabilities; if non-root, behavior is unspecified
- `setForNonRoot` (open to suggestion, I name it based on *runAsNonRoot*): if non-root user, set capabilities; if root, no ops

My goals are: 
1. clarify the root vs non-root cases 
2. clarify how the new field work with existing ones 
3. hide linux details

Reply:
Some New thoughts after talked with Vinayak today.

Option 1: Reuse `add` and `drop`
- The capabilities will be added to ambient set, besides the current capability sets. Thus it will work for non-root user case without file capability. 
- It will not break existing scenarios.

Option 2: Add new field - `ambient`
- Users need to understand all those Linux capability concept: https://lwn.net/Articles/636533/
- We need to enhance the `add` and `drop` spec by saying permitted, inheritable, bounding and effective set.

Option 3: Add new field - `addForNonRoot`
- `add` is for any process(parent or child) with root user
- `drop` is for any process(parent or child) with root user
- `addForNonRoot` is for any process(parent or child) with non-root user